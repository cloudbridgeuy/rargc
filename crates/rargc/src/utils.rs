use convert_case::{Boundary, Converter, Pattern};

/// Convert a string to cobol case
pub fn to_cobol_case(value: &str) -> String {
    Converter::new()
        .set_pattern(Pattern::Uppercase)
        .set_delim("-")
        .set_boundaries(&[Boundary::Underscore, Boundary::LowerUpper, Boundary::Hyphen])
        .convert(value)
}

/// Convert a string from snake_case to camelCase
pub fn hyphens_to_underscores(name: &str) -> String {
    name.replace('-', "_")
}

/// Escape a string for use in a shell command
pub fn escape_shell_words(value: &str) -> String {
    let mut output = String::new();
    if value.is_empty() {
        return "''".to_string();
    }
    for ch in value.chars() {
        match ch {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '_' | '-' | '.' | ',' | ':' | '/' | '@' => {
                output.push(ch)
            }
            '\n' => output.push_str("'\n'"),
            _ => {
                output.push('\\');
                output.push(ch);
            }
        }
    }
    output
}

/// Is the character a valid choice value terminator
pub fn is_choice_value_terminate(c: char) -> bool {
    c == '|' || c == ']'
}

/// Is the character a valid default value terminator
pub fn is_default_value_terminate(c: char) -> bool {
    c.is_whitespace()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cobol() {
        assert_eq!("FOO-BAR".to_string(), to_cobol_case("fooBar"));
        assert_eq!("FOO-BAR".to_string(), to_cobol_case("foo-bar"));
        assert_eq!("FOO1".to_string(), to_cobol_case("foo1"));
    }
}
