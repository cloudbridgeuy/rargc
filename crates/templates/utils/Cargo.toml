[package]
name = "utils"
version = "0.0.0"
homepage = "https://github.com/cloudbridgeuy/rargc"
description = "Tests utility library"
autobins = false

authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
name = "utils"
path = "src/lib.rs"

[dependencies]



