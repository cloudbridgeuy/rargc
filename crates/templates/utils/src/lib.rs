#[macro_export]
macro_rules! test_template {
    ($template:expr, $description:expr, $json:expr) => {
        let object = Context::from_serialize($json);

        let output =
            match TEMPLATES.render($template, &object.expect("Can't create JSON object")) {
                Ok(o) => o,
                Err(e) => {
                    log::error!("Parsing error(s): {}", e);
                    ::std::process::exit(1);
                }
            };

        insta::with_settings!({
            description => $description,
        }, {
            insta::assert_snapshot!(output)
        });
    }
}
