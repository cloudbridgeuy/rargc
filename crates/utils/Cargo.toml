[package]
name = "utils"
version = "0.0.0"
homepage = "https://github.com/cloudbridgeuy/rargs"
description = "Tests utility library"
autobins = false

authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
name = "utils"
path = "src/lib.rs"

[dependencies]
paste = "1.0.14"
