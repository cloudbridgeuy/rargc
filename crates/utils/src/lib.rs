pub extern crate paste;

#[macro_export]
macro_rules! test_template {
    ($template:expr, $description:expr, $json:expr) => {
        let object = Context::from_serialize($json);

        let output =
            match TEMPLATES.render($template, &object.expect("Can't create JSON object")) {
                Ok(o) => o,
                Err(e) => {
                    log::error!("Parsing error(s): {}", e);
                    ::std::process::exit(1);
                }
            };

        insta::with_settings!({
            description => $description,
        }, {
            insta::assert_snapshot!(output)
        });
    }
}

#[macro_export]
macro_rules! test_script {
    ($script:expr, $stream:expr, $options:expr) => {
        let script_name = $script.to_string();
        let options = $options.to_string();
        let stream = $stream.to_string();
        let description = format!("[{stream}] {script_name} {options}");
        let output_dir = "./examples/output".to_string();
        let src = format!("./examples/{script_name}");
        let dst = format!("./examples/output/{script_name}");

        log::info!("{}", description);
        let output = Command::new(&dst)
            .args(options.split(" ").collect::<Vec<&str>>())
            .output()
            .unwrap_or_else(|e| panic!("failed to execute the '{dst}' script with error: {e}"));
        insta::with_settings!({
            description => description,
        }, {
            insta::assert_snapshot!(String::from_utf8(match $stream {
                "stdout" => output.stdout,
                "stderr" => output.stderr,
                _ => panic!("invalid stream"),
            }).unwrap_or_else(|e| panic!("failed to execute the '{dst}' script with error: {e}")))
        });
    };
}

#[macro_export]
macro_rules! build_script {
    ($script:expr) => {
        let script_name = $script.to_string();
        let output_dir = "./examples/output".to_string();
        let src = format!("./examples/{script_name}");
        let dst = format!("./examples/output/{script_name}");

        log::info!("Build ./examples/output/minimal.rs");
        let output = Command::new("cargo")
            .args([
                "run",
                "--bin",
                "rargc",
                "--release",
                "--",
                "build",
                "-d",
                &output_dir,
                &src,
            ])
            .output()
            .unwrap_or_else(|e| panic!("failed to build rargc script with error: {e}"));

        println!("stdout: {}", String::from_utf8(output.stdout).unwrap());
        println!("stderr: {}", String::from_utf8(output.stderr).unwrap());

        log::info!("Check that the {dst} file exists");
        assert!(std::path::Path::new(&dst).exists());

        // Read the `dst` file into a variable of type `&str`
        let dst = std::fs::read_to_string(&dst)
            .unwrap_or_else(|e| panic!("failed to read the '{dst}' script with error: {e}"));

        insta::with_settings!({
            description => script_name,
        }, {
            insta::assert_snapshot!(dst)
        });
    };
}
