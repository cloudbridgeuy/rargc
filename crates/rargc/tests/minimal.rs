use std::process::Command;
use test_log::test;

use utils::{build_script, test_script};

#[test]
fn test_minimal() {
    let script_name = "minimal.sh".to_string();

    let commands: Vec<(&str, &str)> = vec![
        ("stderr", ""),
        ("stdout", "--help"),
        ("stderr", "-f"),
        ("stdout", "source -f"),
        ("stdout", "source target -f"),
    ];

    build_script!(script_name);

    for (stream, options) in commands {
        test_script!(script_name, stream, options);
    }
}
