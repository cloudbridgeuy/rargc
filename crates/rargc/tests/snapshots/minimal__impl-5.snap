---
source: crates/rargc/tests/minimal.rs
description: "Run the minimal.sh script with '[\"source\", \"target\", \"-f\"]'"
expression: "String::from_utf8(match \"stdout\" {\n            \"stdout\" => output.stdout,\n            \"stderr\" => output.stderr,\n            _ => panic!(\"invalid stream\"),\n        }).unwrap_or_else(|e|\n        panic!(\"failed to execute the '{dst}' script with error: {e}\"))"
---
# this file is located in './crates/rargc/examples/output.sh'
# you can edit it freely and regenerate (it will not be overwritten)
args:
- ${args[--force]} = 1
- ${args[source]} = source
- ${args[target]} = target

